Tag evolution

india-kerle opened the pull request:
I've run both main scripts (clustering and propagating labels via a network approach and via an embedding approach) and checked the output of both. I've also added a json decoder so we can save np.arrays to S3. I've added a readme with a sentence or two on a few high level observations but would need to be fleshed out a bit more.  

The main files to look into are:
`ai_genomics/analysis/tag_evolution/tag_evolution_embeds.py`
`ai_genomics/analysis/tag_evolution/tag_evolution_network.py`

An issue that's come out is that filtering the OpenAlex entities based on their ID keys takes ages as its a massive file - i think we'll need to write a script that takes a sample then saves the sample then we can load it (or something along these lines) -can do in a separate PR.

Commit:
added patents wranlger class to clean patents

Commit:
Merge branch 'dev' of github.com:nestauk/ai_genomics into patents_db_tags

Commit:
updated class description

Commit:
added tag evolution and vectoriser script

Commit:
rename tag evolution using embeds, add tag evolution network approach, add pec utils, update data_getters

Commit:
updated tag evolution scripts

Commit:
remove trailing whitespace

Commit:
removed patents wrangler

Commit:
update requirements.txt

Commit:
updated utils

Commit:
updated utils

Commit:
removed oa for now

Commit:
added methods desc

Commit:
update readme

Commit:
updated readme and added scripts

Commit:
added readme and approaches

Commit:
remove sys

Commit:
Merge branch 'dev' of github.com:nestauk/ai_genomics into tag_evolution

Commit:
added OA tags, fixed bugs

Commit:
updated high level observations

Commit:
get rid of sys

Juan-Mateos commented on ai_genomics/analysis/tag_evolution/README.md:
```diff
@@ -0,0 +1,35 @@
+
+The scripts in this directory focus on clustering AI Genomics tags across all datasets over time in two approaches: 1) a semantic approach and 2) a tag co-occurance network based approach.
+
+In both instances, DBpedia entities are cumulatively 'timesliced' based on founded, publication or granted dates across the data sources. Cluster names are propagated based on a minimum jaccard similarity score of at least 0.5 for given clusters t and t+1.
```
is this minimum jaccard similarity between clusters in different periods?

india-kerle replied:
yes !

Juan-Mateos commented on ai_genomics/analysis/tag_evolution/tag_evolution_embeds.py:
```diff
@@ -0,0 +1,244 @@
+        .reset_index(drop=True)
+        .rename(columns={date_col: "date", id_col: "id"})
+    )
+    data["date"] = pd.to_datetime(data["date"]).dt.date
```
this could be chained to the previous with `.assign(date = lambda df: pd.to_datetime(df["date"]).dt.date)`
]

and then you could just return the whole thing without assigning to `data` (which might by the way change the `data` df outside the function)

Juan-Mateos commented on ai_genomics/analysis/tag_evolution/tag_evolution_embeds.py:
```diff
@@ -0,0 +1,244 @@
+reducer = umap.UMAP()
+
+# load entities
+patent_ents, crunchbase_ents, gtr_ents, oa_ents = (
```
Why not read these after `if __name__ == "__main__":`?

Juan-Mateos commented on ai_genomics/analysis/tag_evolution/tag_evolution_embeds.py:
```diff
@@ -0,0 +1,244 @@
+
+
+def propagate_labels(
+    timeslice_x, timeslice_y, min_jaccard_score: float = 0.7,
```
typing and docstrings would help here

india-kerle replied:
added - thanks

Juan-Mateos commented on ai_genomics/analysis/tag_evolution/tag_evolution_embeds.py:
```diff
@@ -0,0 +1,244 @@
+        dict([(i[0], i[1]) for i in sorted_perm_dists]).items()
+    )
+
+    while len(sorted_perm_dists_clusts) > 0:
```
I struggled to follow this function because it doesn't return anything. Are you modifying `timeslice_x` and `timeslice_y` in place? I think it would be nicer to return new modified objects. 

It might be easier to understand the logic if you change the name of arguments to e.g. `timeslice_t` and `timeslice_t_plus_1` or something like that.

india-kerle replied:
good idea- i've changed the variable names to be easier to understand and i've commented each line of code with a description of what's happening so hopefully its easier to follow

Juan-Mateos commented on ai_genomics/analysis/tag_evolution/tag_evolution_embeds.py:
```diff
@@ -0,0 +1,244 @@
+        },
+        "outputs/analysis/tag_evolution/dbpedia_clusters_timeslice_embed.json",
+    )
+    logger.info("saved dbpedia entities and propagated entity clusters.")
```
Where are you saving the timeslices with propagated names? I thought those were stored in `timeslice_x` and `timeline_y`? Or are the changes to that propagating to `ents_per_date_clusts` somehow? I would try to make that a bit clearer.

india-kerle replied:
the changes happen in place to ents_per_date_clusts - I've added comments so its easier to follow

Juan-Mateos commented on ai_genomics/analysis/tag_evolution/tag_evolution_network.py:
```diff
@@ -0,0 +1,163 @@
+"""This script clusters an AI genomics DBepdia coocurrance network at successive 
```
As discussed I am ignoring this script.

Juan-Mateos commented on ai_genomics/utils/network_time_series.py:
```diff
@@ -0,0 +1,241 @@
+"""
+Util functions to create subgraphs by timeslice, cluster timeslices and propagate
+cluster labels across timeslices to make communities comparable across time.
```
I didn't review this

Juan-Mateos reviewed (commented):
In order to install the required packages I had to put `boto3 and s3fs=2022.5.0` in `environment.yaml` and run `conda update -f environment.yaml`.

Having done this, I reproduced the `tag_evolution_embeds.py` script. 

I have left a couple of minor comments. In particular, it would be useful to clarify where are the progagated labels stored and saved.

Commit:
deal with merge conflict

Commit:
deleted network label propagation and updated george's utils

Commit:
finished reduce docstrings

Commit:
address juans comments and refactor temporal clustering

Commit:
address juans comments and refactor temporal clustering

Commit:
fix logger info

Commit:
change reduce embedding funct

Commit:
debugged scripts

Commit:
remove sys

Commit:
address juans pr

india-kerle commented:
thanks for the PR @Juan-Mateos - i've refactored the code to use some of george's utils + added to them. I've also added more thorough docstrings + comments in the code so the logic is easier to follow - do let me know how you get on with these changes.

Commit:
updated requirements

Juan-Mateos reviewed (approved):
Good to merge, thank you!
